File: public/code/popular__ttm_squeeze.py
from common import make_signal, ema, sma, rsi, macd, atr, bollinger_bands, vwap, swing_highs, swing_lows
import math


def _sign(x):
    return (x > 0) - (x < 0)


def mk(c, i, side, a, conf, reason, m=2):
    cur = c[i]
    if side == "long":
        sl = cur['close'] - m * a[i]
        r = cur['close'] - sl
        return make_signal(signal="long", entry=cur['close'], stop_loss=sl, take_profit=[cur['close'] + r * 1.5, cur['close'] + r * 2.5, cur['close'] + r * 4], confidence=conf, reason=reason)
    sl = cur['close'] + m * a[i]
    r = sl - cur['close']
    return make_signal(signal="short", entry=cur['close'], stop_loss=sl, take_profit=[cur['close'] - r * 1.5, cur['close'] - r * 2.5, cur['close'] - r * 4], confidence=conf, reason=reason)
def hma(values, period):
    def wma(arr, p):
        out = []
        for i in range(0, len(arr)):
            if i < p - 1:
                out.append(arr[i])
                continue
            num = 0
            den = 0
            for k in range(0, p):
                w = p - k
                num += arr[i - k] * w
                den += w
            out.append(num / den)
        return out
    half = max(1, math.floor(period / 2))
    sqrtP = max(1, round(math.sqrt(period)))
    w1 = wma(values, half)
    w2 = wma(values, period)
    raw = [2 * v - w2[i] for i, v in enumerate(w1)]
    return wma(raw, sqrtP)
def keltner(c, p, mult):
    closes = [x['close'] for x in c]
    mid = ema(closes, p)
    a = atr(c, p)
    return {'mid': mid, 'upper': [v + mult * a[i] for i, v in enumerate(mid)], 'lower': [v - mult * a[i] for i, v in enumerate(mid)]}
def ttmSqueeze(c):
    if len(c) < 40:
        return make_signal(reason="Insufficient data")
    closes = [x['close'] for x in c]
    bb = bollingerBands(closes, 20, 2)
    kc = keltner(c, 20, 1.5)
    a = atr(c, 14)
    i = len(c) - 1
    squeezeOn = lambda idx: bb['lower'][idx] > kc['lower'][idx]  and  bb['upper'][idx] < kc['upper'][idx]
    # Momentum: close - midpoint donchian/sma
    mom = [((0 if idx < 20 else v - (max([x['high'] for x in c[idx - 20: idx]]) + min([x['low'] for x in c[idx - 20: idx]]) + sma(closes, 20)[idx]) / 3)) for idx, v in enumerate(closes)]
    # Squeeze biraz önce vardı, şimdi bitti (fire) + momentum yönü
    if squeezeOn(i - 1)  and  not squeezeOn(i)  and  mom[i] > 0:
        return mk(c, i, "long", a, 0.74, "TTM Squeeze fired long (momentum up)")
    if squeezeOn(i - 1)  and  not squeezeOn(i)  and  mom[i] < 0:
        return mk(c, i, "short", a, 0.74, "TTM Squeeze fired short (momentum down)")
    return make_signal(reason=("Squeeze on (waiting for release)" if squeezeOn(i) else "No squeeze release"))

File: public/code/popular__ttm_squeeze.ts
import { Candle, Signal, makeSignal, sma, ema, rsi, atr, bollingerBands, macd } from "../indicators";

const mk = (c: Candle[], i: number, side: "long" | "short", a: number[], conf: number, reason: string, m = 2): Signal => {
  const cur = c[i];
  if (side === "long") { const sl = cur.close - m * a[i], r = cur.close - sl; return makeSignal({ signal: "long", entry: cur.close, stop_loss: sl, take_profit: [cur.close + r * 1.5, cur.close + r * 2.5, cur.close + r * 4], confidence: conf, reason }); }
  const sl = cur.close + m * a[i], r = sl - cur.close; return makeSignal({ signal: "short", entry: cur.close, stop_loss: sl, take_profit: [cur.close - r * 1.5, cur.close - r * 2.5, cur.close - r * 4], confidence: conf, reason });
};
function hma(values: number[], period: number): number[] {
  const wma = (arr: number[], p: number): number[] => {
    const out: number[] = [];
    for (let i = 0; i < arr.length; i++) {
      if (i < p - 1) { out.push(arr[i]); continue; }
      let num = 0, den = 0;
      for (let k = 0; k < p; k++) { const w = p - k; num += arr[i - k] * w; den += w; }
      out.push(num / den);
    }
    return out;
  };
  const half = Math.max(1, Math.floor(period / 2));
  const sqrtP = Math.max(1, Math.round(Math.sqrt(period)));
  const w1 = wma(values, half), w2 = wma(values, period);
  const raw = w1.map((v, i) => 2 * v - w2[i]);
  return wma(raw, sqrtP);
}
function keltner(c: Candle[], p: number, mult: number) {
  const closes = c.map((x) => x.close), mid = ema(closes, p), a = atr(c, p);
  return { mid, upper: mid.map((v, i) => v + mult * a[i]), lower: mid.map((v, i) => v - mult * a[i]) };
}

export function ttmSqueeze(c: Candle[]): Signal {
  if (c.length < 40) return makeSignal({ reason: "Insufficient data" });
  const closes = c.map((x) => x.close), bb = bollingerBands(closes, 20, 2), kc = keltner(c, 20, 1.5), a = atr(c, 14), i = c.length - 1;
  const squeezeOn = (idx: number) => bb.lower[idx] > kc.lower[idx] && bb.upper[idx] < kc.upper[idx];
  // Momentum: close - midpoint donchian/sma
  const mom = closes.map((v, idx) => (idx < 20 ? 0 : v - (Math.max(...c.slice(idx - 20, idx).map((x) => x.high)) + Math.min(...c.slice(idx - 20, idx).map((x) => x.low)) + sma(closes, 20)[idx]) / 3));
  // Squeeze biraz önce vardı, şimdi bitti (fire) + momentum yönü
  if (squeezeOn(i - 1) && !squeezeOn(i) && mom[i] > 0) return mk(c, i, "long", a, 0.74, "TTM Squeeze fired long (momentum up)");
  if (squeezeOn(i - 1) && !squeezeOn(i) && mom[i] < 0) return mk(c, i, "short", a, 0.74, "TTM Squeeze fired short (momentum down)");
  return makeSignal({ reason: squeezeOn(i) ? "Squeeze on (waiting for release)" : "No squeeze release" });
}
